python/lsst/daf/butler/core/location.py

```diff
@@ -54,6 +55,81 @@ def path(self):
         """
         return os.path.join(self._datastoreRoot, self._uri.path.lstrip("/"))
 
+    @property
+    def component_path(self):
+        """Path corresponding to location of component of composite.
+
+        This is used in cases where a composite is broken into separate
+        files, combining the root file path with the component name.
+        When reading a file from the URI, it is possible that the file
+        will not exist if the composite was written as a single entity
+        and the URI fragment is referring to a location inside that entity.
+
+        This path includes the root of the `Datastore`. Returns the empty
+        string if no fragment has been specified as part of the URI.
+        """
+        if not self.fragment:
+            return ''
+
+        # We need to insert the fragment before the file extension because
+        # some I/O libraries insist on a particular extension.
+        filepath, extension = os.path.splitext(self.path)
+        return "{}#{}{}".format(filepath, self.fragment, extension)
+
+    @property
+    def fragment(self):
+        """URI fragment associated with this location.
+        """
+        return self._uri.fragment
+
+    def paths(self):
+        """Return path and component path.
+
+        Returns
+        -------
+        path : `tuple`
+            Tuple of the path and the component path.
+            The component path can be an empty string.
+        """
+        return (self.path, self.component_path)
+
+    def preferredPath(self):
+        """Returns component path, if set, else the path.
+
+        Returns
+        -------
+            Path for component in composite if set, else the default path.
+        """
+        return self.component_path if self.component_path else self.path
```

Comment:
we never said that user configuration would control how components are split up yet, and we don't yet have templating in config files. When templating is added to datastore configurations we will likely have to revisit this. `Location` has to know the component storage scheme at the moment because `Location` is currently used by other parts of datastore to determine whether a file exists or not independent of formatter. Once we switch to datastore having to map dataIDs to filenames in a database table `Location` won't need to know any more.